### PR TITLE
Add op read_file and decode_jpeg

### PR DESCRIPTION
**Pull request by LielinJiang**
### PR types

New features
### PR changes

APIs
### Describe

Add op read_file and decode_jpeg

file_path -> gpu tensor的时间比较（单位：s/1000次）
image

### CODE REVIEW

File: paddle/fluid/operators/decode_jpeg_op.cc [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+class DecodeJpegOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X",
+             "A one dimensional uint8 tensor containing the raw bytes "
+             "of the JPEG image. It is a tensor with rank 1.");
+    AddOutput("Out", "The output tensor of DecodeJpeg op");
+    AddComment(R"DOC(
+This operator decode a JPEG image into a 3 dimensional RGB Tensor.
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cc]: 下面的注释是支持灰度图的吧，这里需要正确的comments

File: paddle/fluid/operators/decode_jpeg_op.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+class DecodeJpegOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("X"), "Input", "X", "DecodeJpeg");
+    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out", "DecodeJpeg");
+
+    auto out_dims = std::vector<int>(1, -1);
+    ctx->SetOutputDim("Out", framework::make_ddim(out_dims));
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cc:60]: out_dims的rank是3吧，这里的shape不正确， 依据mode， 是 {1, -1, -1} 或 {3，-1， -1}

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/platform/dynload/nvjpeg.h"
+#include "paddle/fluid/platform/enforce.h"
+
+namespace paddle {
+namespace operators {
+
+static nvjpegHandle_t nvjpeg_handle = nullptr;
+
+void init_nvjpegImage(nvjpegImage_t* img) {
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cu]: InitNvjpegImage

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    PADDLE_ENFORCE_EQ(
+        info_status, NVJPEG_STATUS_SUCCESS,
+        platform::errors::Fatal("nvjpegGetImageInfo failed: ", info_status));
+
+    int width = widths[0];
+    int height = heights[0];
+
+    nvjpegOutputFormat_t outputFormat;
+    int outputComponents;
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cu]: outputFormat -> output_format
outputComponents -> output_components

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+      outputComponents = 3;
+    } else {
+      platform::dynload::nvjpegJpegStateDestroy(nvjpeg_state);
+      PADDLE_ENFORCE_EQ(
+          false, true,
+          platform::errors::Fatal(
+              "The provided mode is not supported for JPEG files on GPU"));
+    }
+
+    nvjpegImage_t outImage;
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cu]: outImage -> out_image

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    T* data = out->mutable_data<T>(ctx.GetPlace());
+
+    for (int c = 0; c < outputComponents; c++) {
+      outImage.channel[c] = data + c * sz;
+      outImage.pitch[c] = width;
+    }
+
+    nvjpegStatus_t decode_status = platform::dynload::nvjpegDecode(
+        nvjpeg_handle, nvjpeg_state, x_data, x->numel(), outputFormat,
+        &outImage, stream);
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cu]: Maybe need another cuda stream

File: paddle/fluid/operators/read_file_op.cc [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    input.read(image_data.data(), file_size);
+
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+    std::vector<int64_t> out_shape = {file_size};
+    out->Resize(framework::make_ddim(out_shape));
+
+    uint8_t* data = out->mutable_data<T>(ctx.GetPlace());
+    std::memcpy(data, reinterpret_cast<uint8_t*>(image_data.data()), file_size);
```
Comment (qingqing01) [inline paddle/fluid/operators/read_file_op.cc]: 看下能否直接读取到out里，去掉拷贝

File: paddle/fluid/operators/read_file_op.cc [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+class ReadFileOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE_EQ(ctx->HasOutput("Out"), true,
+                      platform::errors::InvalidArgument(
+                          "Output(Out) of ReadFileOp is null."));
+
+    auto out_dims = std::vector<int>(1, -1);
```
Comment (qingqing01) [inline paddle/fluid/operators/read_file_op.cc:58]: rank是1的话，这里的out_dims是 {-1}, （1，-1)代表的rank是2

File: python/paddle/tests/test_read_file.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline python/paddle/tests/test_read_file.py]: 21

File: python/paddle/vision/ops.py [hunk 10]
```diff
@@ -782,3 +785,95 @@ def forward(self, x, offset, mask=None):
...
+    out = helper.create_variable_for_type_inference('uint8')
+    helper.append_op(
+        type="read_file", inputs=inputs, attrs=attrs, outputs={"Out": out})
+
+    return out
+
+
+def decode_jpeg(x, mode='unchanged', name=None):
+    """
+    Decodes a JPEG image into a 3 dimensional RGB Tensor.
```
Comment (qingqing01) [inline python/paddle/vision/ops.py]: Decodes a JPEG image into a 3 dimensional RGB Tensor or 1 dimensional Gray Tensor

File: paddle/fluid/operators/decode_jpeg_op.cc (hunk 1 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cc]: done, thanks

File: paddle/fluid/operators/decode_jpeg_op.cc (hunk 2 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cc:60]: done, thanks

File: paddle/fluid/operators/decode_jpeg_op.cu (hunk 3 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cu]: done, thanks

File: paddle/fluid/operators/decode_jpeg_op.cu (hunk 4 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cu]: done, thanks

File: paddle/fluid/operators/decode_jpeg_op.cu (hunk 5 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cu]: done, thanks

File: paddle/fluid/operators/decode_jpeg_op.cu (hunk 6 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cu]: done, thanks

File: paddle/fluid/operators/read_file_op.cc (hunk 7 above)
Comment (LielinJiang) [reply paddle/fluid/operators/read_file_op.cc]: done, thanks

File: python/paddle/tests/test_read_file.py (hunk 9 above)
Comment (LielinJiang) [reply python/paddle/tests/test_read_file.py]: done, thanks

File: python/paddle/vision/ops.py (hunk 10 above)
Comment (LielinJiang) [reply python/paddle/vision/ops.py]: done, thanks

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    T* data = out->mutable_data<T>(ctx.GetPlace());
+
+    for (int c = 0; c < output_components; c++) {
+      out_image.channel[c] = data + c * sz;
+      out_image.pitch[c] = width;
+    }
+
+    nvjpegStatus_t decode_status = platform::dynload::nvjpegDecode(
+        nvjpeg_handle, nvjpeg_state, x_data, x->numel(), output_format,
+        &out_image, nvjpeg_stream);
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cu:128]: 需要调用nvjpegJpegStateDestroy释放nvjpeg_state

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+    // create nvjpeg stream
+    if (nvjpeg_stream == nullptr) {
+      cudaStreamCreateWithFlags(&nvjpeg_stream, cudaStreamNonBlocking);
+    }
+
+    int sz = widths[0] * heights[0];
+
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+    std::vector<int64_t> out_shape = {output_components, height, width};
```
Comment (lyuwenyu) [inline paddle/fluid/operators/decode_jpeg_op.cu:116]: int64_t -> uint64_t？

File: paddle/fluid/operators/read_file_op.cc [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    auto filename = ctx.Attr<std::string>("filename");
+
+    std::ifstream input(filename.c_str(),
+                        std::ios::in | std::ios::binary | std::ios::ate);
+    std::streamsize file_size = input.tellg();
+
+    input.seekg(0, std::ios::beg);
+
+    auto* out = ctx.Output<framework::LoDTensor>("Out");
+    std::vector<int64_t> out_shape = {file_size};
```
Comment (lyuwenyu) [inline paddle/fluid/operators/read_file_op.cc:40]: as mentioned above

File: paddle/fluid/operators/decode_jpeg_op.cu [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/platform/dynload/nvjpeg.h"
+#include "paddle/fluid/platform/enforce.h"
+#include "paddle/fluid/platform/stream/cuda_stream.h"
+
+namespace paddle {
+namespace operators {
+
+static cudaStream_t nvjpeg_stream = nullptr;
+static nvjpegHandle_t nvjpeg_handle = nullptr;
```
Comment (qingqing01) [inline paddle/fluid/operators/decode_jpeg_op.cu:27]: 后续需要考虑如何管理这种static变量

**Review by qingqing01 — APPROVED** [review]

File: paddle/fluid/operators/decode_jpeg_op.cu (hunk 12 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cu:116]: function make_dims only accept type int64_t

File: paddle/fluid/operators/decode_jpeg_op.cu (hunk 14 above)
Comment (LielinJiang) [reply paddle/fluid/operators/decode_jpeg_op.cu:27]: 好的

**Pull request merged**